src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -381,6 +381,41 @@
    "UPDATE factsets SET hash=md5(factsets.id::text)::bytea WHERE hash is NULL"
    "ALTER TABLE factsets ALTER COLUMN hash SET NOT NULL"))
 
+(defn add-support-for-historical-catalogs []
+  (jdbc/do-commands
+   "ALTER TABLE catalog_resources RENAME TO catalog_resources_tmp"
+   ;; CREATE certnames and catalog_resources transform tables
+   "CREATE TABLE catalog_resources (LIKE catalog_resources_tmp INCLUDING ALL)"
+   "CREATE TABLE latest_catalogs (catalog_id BIGINT NOT NULL REFERENCES catalogs(id) ON DELETE CASCADE, certname_id BIGINT PRIMARY KEY REFERENCES certnames(id) ON DELETE CASCADE)"
+   "ALTER TABLE catalog_resources DROP COLUMN catalog_id"
+   "ALTER TABLE catalog_resources ADD COLUMN certname_id BIGINT NOT NULL REFERENCES certnames(id) ON DELETE CASCADE"
+   "ALTER TABLE catalog_resources ADD PRIMARY KEY (certname_id, type, title)"
+
+   (str "INSERT INTO latest_catalogs"
+        "  (catalog_id, certname_id)"
+        "  SELECT catalogs.id, certnames.id"
+        "  FROM certnames"
+        "  INNER JOIN catalogs ON catalogs.certname = certnames.certname")
+
+   (str "INSERT INTO catalog_resources"
+        "  (certname_id, resource, tags, type, title, exported, file, line)"
+        "  SELECT latest_catalogs.certname_id, cr.resource, cr.tags, cr.type, cr.title, cr.exported, cr.file, cr.line"
+        "  FROM catalog_resources_tmp cr"
+        "  INNER JOIN latest_catalogs ON cr.catalog_id = latest_catalogs.catalog_id")
+
+   "DROP TABLE catalog_resources_tmp"
+
+   "ALTER TABLE catalog_resources ADD CONSTRAINT catalog_resources_resource_fkey FOREIGN KEY (resource) REFERENCES resource_params_cache(resource) ON DELETE CASCADE"
+
+   "ALTER TABLE catalogs DROP CONSTRAINT catalogs_certname_key"
+   "ALTER TABLE catalogs DROP CONSTRAINT catalogs_hash_key"
+   "ALTER TABLE catalogs ADD COLUMN edges JSONB DEFAULT NULL"
+   "ALTER TABLE catalogs ADD COLUMN resources JSONB DEFAULT NULL"
+   "ALTER TABLE catalogs ADD COLUMN catalog_uuid UUID DEFAULT NULL"
+   "ALTER TABLE reports ADD COLUMN catalog_uuid UUID DEFAULT NULL"
+
+   "CREATE INDEX reports_catalog_uuid ON reports USING btree (catalog_uuid)"))
```

Comment:
On closer inspection, I think we want to keep this one for garbage collection, otherwise we'll have to change the GC query to cast to text. Maybe we want 2 indexes on this column?